File: bot/src/main/java/com/ryderbelserion/discord/api/commands/CommandHandler.java
package com.ryderbelserion.discord.api.commands;

import com.ryderbelserion.discord.api.commands.interfaces.CommandFlow;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import org.jspecify.annotations.NonNull;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class CommandHandler implements CommandFlow {

    private final Map<String, CommandEngine> commands = new HashMap<>();

    private JDA jda;

    public void setJda(@NonNull final JDA jda) {
        this.jda = jda;
    }

    @Override
    public void addCommands(@NonNull final List<CommandEngine> commands) {
        this.jda.updateCommands().addCommands(commands.stream()
                .map(CommandEngine::getCommandData)
                .collect(Collectors.toSet())).queue();

        for (final CommandEngine engine : commands) { // add to cache
            this.commands.put(engine.getName(), engine);
        }
    }

    @Override
    public void addStaticGuildCommands(@NonNull final Guild guild, @NonNull final List<CommandEngine> commands) { // statically generated guild commands, i.e. on startup only, hard coded.
        addGuildCommands(guild, commands.stream().map(CommandEngine::getCommandData).toList());

        for (final CommandEngine engine : commands) { // add to cache
            this.commands.put(engine.getName(), engine);
        }
    }

    @Override
    public void addGuildCommands(@NonNull final Guild guild, @NonNull final List<CommandData> commands) { // dynamically generated guild commands, i.e. yml files
        guild.updateCommands().addCommands(commands).queue();
    }

    @Override
    public void purgeGuildCommands(@NonNull final Guild guild) {
        guild.updateCommands().queue();
    }

    @Override
    public void purgeCommands() {
        this.jda.updateCommands().queue();
    }

    public @NonNull final CommandEngine getCommand(@NonNull final String name) {
        return this.commands.get(name);
    }

    public final boolean hasCommand(@NonNull final String name) {
        return this.commands.containsKey(name);
    }
}

File: src/main/kotlin/com/ryderbelserion/api/builders/commands/CommandHandler.kt
package com.ryderbelserion.api.builders.commands

import com.ryderbelserion.api.builders.commands.interfaces.CommandFlow
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.interactions.commands.OptionType
import net.dv8tion.jda.api.interactions.commands.build.OptionData

class CommandHandler(private val jda: JDA) : CommandFlow {

    private lateinit var guild: Guild

    override fun addCommand(engine: CommandEngine) {
        this.jda.upsertCommand(engine.name, engine.description).queue()
    }

    override fun addCommand(engine: CommandEngine, optionData: OptionData) {
        this.jda.upsertCommand(engine.name, engine.description).addOptions(optionData).queue()
    }

    override fun addCommand(engine: CommandEngine, optionDataList: MutableList<OptionData?>) {
        this.jda.upsertCommand(engine.name, engine.description).addOptions(optionDataList).queue()
    }

    override fun addCommand(engine: CommandEngine, type: OptionType, name: String, description: String) {
        this.jda.upsertCommand(engine.name, engine.description).addOption(type, name, description).queue()
    }

    override fun addGuildCommand(guild: Guild, engine: CommandEngine) {
        guild.upsertCommand(engine.name, engine.description).queue()
    }

    override fun addGuildCommand(guild: Guild, engine: CommandEngine, optionData: OptionData) {
        guild.upsertCommand(engine.name, engine.description).addOptions(optionData).queue()
    }

    override fun addGuildCommand(guild: Guild, engine: CommandEngine, optionDataList: MutableList<OptionData?>) {
        guild.upsertCommand(engine.name, engine.description).addOptions(optionDataList).queue()
    }

    override fun addGuildCommand(
        guild: Guild,
        engine: CommandEngine,
        type: OptionType,
        name: String,
        description: String
    ) {
        guild.upsertCommand(engine.name, engine.description).addOption(type, name, description).queue()
    }

    override fun addGuildCommands(guild: Guild, vararg engines: CommandEngine) {
        for (engine in engines) {
            addGuildCommand(guild, engine)
        }
    }

    override fun addCommands(vararg engines: CommandEngine) {
        for (engine in engines) {
            addCommand(engine)
        }
    }

    override fun purgeGuildCommands(guild: Guild) {
        guild.updateCommands().queue()
    }

    override fun purgeGlobalCommands() {
        this.jda.updateCommands().queue()
    }

    fun guild(guild: Guild): CommandHandler {
        this.guild = guild

        return this
    }
}